crates/subspace-farmer/src/commitments/commitment_databases.rs

```diff
@@ -95,62 +103,125 @@ impl CommitmentDatabases {
             .collect()
     }
 
-    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<&Arc<DbEntry>> {
-        self.databases.peek(salt)
+    pub(super) fn get_db_entry(&self, salt: &Salt) -> Option<Arc<DbEntry>> {
+        return match salt {
+            _ if &self.first_db.read().as_ref().unwrap().salt == salt => {
+                self.first_db.read().clone()
+            }
+            _ if &self.second_db.read().as_ref().unwrap().salt == salt => {
+                self.second_db.read().clone()
+            }
+            _ => None,
+        };
     }
 
     /// Returns `Ok(None)` if entry for this salt already exists.
     pub(super) fn create_db_entry(
-        &mut self,
+        &self,
         salt: Salt,
+        current_salt: Salt,
     ) -> Result<Option<CreateDbEntryResult>, CommitmentError> {
-        if self.databases.contains(&salt) {
-            return Ok(None);
-        }
+        match salt {
+            _ if self.first_db.read().as_ref().is_some()
+                && self.first_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ if self.second_db.read().as_ref().is_some()
+                && self.second_db.read().as_ref().unwrap().salt == salt =>
+            {
+                return Ok(None)
+            }
+            _ => {} // don't do anything yet if salt does not exist
+        };
 
+        // create the new db_entry
         let db_entry = Arc::new(DbEntry {
             salt,
             db: Mutex::new(None),
         });
         let mut removed_entry_salt = None;
+        let old_db_entry;
 
-        let old_db_entry = if self.databases.len() >= COMMITMENTS_CACHE_SIZE {
-            self.databases.pop_lru().map(|(_salt, db_entry)| db_entry)
-        } else {
-            None
+        // delete the old database only, and don't touch the other to prevent data races
+        match current_salt {
+            // if the `first` DB is not the current one, replace it
+            _ if self.first_db.read().as_ref().is_none() => {
+                old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ if self.second_db.read().as_ref().is_none() => {
+                old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ if self.first_db.read().as_ref().is_some()
+                && self.first_db.read().as_ref().unwrap().salt != current_salt =>
+            {
+                old_db_entry = self.first_db.write().replace(Arc::clone(&db_entry));
+            }
+            // if the `second` DB is not the current one, replace it
+            _ if self.second_db.read().as_ref().is_some()
+                && self.second_db.read().as_ref().unwrap().salt != current_salt =>
+            {
+                old_db_entry = self.second_db.write().replace(Arc::clone(&db_entry));
+            }
+            _ => unreachable!(),
```

Comment:
Ok, here is what's going on:
if `first_db` is none, we simply replace it with the new db.
if `second_db` is none, we simply replace it with the new db.
if `first_db` is not none, and its salt is not equal to the current salt (if it is not the active database), we replace it with the new db.
if `second_db is not none, and if its not the active database, we replace it with the new db.

And there should be no other case. 
However, this 4-way pattern matching seems overly repetitive for me, and I have a feeling that this could be reduced to 2 patterns. I just don't know how.

Reply:
> I'm not going to analyze deeper for now because this is already problematic. You check with `.read()`, release the lock and then do `.write()`, opening window for race conditions where something can modify contents between `.read()` and `.write()` calls.

I don't think it is problematic. Yes, something else can modify the contents in between the read/write locks. But it won't. AFAIK, nothing is modifying the inactive commitments database, it is already the **old** one. So basically, this race condition will never happen in our protocol (unless there is a detail I'm missing).